File: solutions/prolog/bob/1/bob.pl
% Dedicated to Shree DR.MDD — my eternal inspiration

:- use_module(library(dcg/basics)).

any_chars --> [] | [_], any_chars.

non_small_char --> [X], { nonvar(X), \+ code_type(X, lower) }.
non_small_seq --> non_small_char, non_small_seq.
non_small_seq --> [].

capital_char --> [X], { nonvar(X), code_type(X, upper) }.

quiet_line --> blanks, eol.
inquiry_line --> any_chars, "?", blanks, eol.
loud_line --> non_small_seq, capital_char, non_small_seq, eol.

pattern_match(Str, Rule) :-
    string_codes(Str, Codes),
    phrase(Rule, Codes).

hey(Text, "Fine. Be that way!") :- pattern_match(Text, quiet_line), !.
hey(Text, "Calm down, I know what I'm doing!") :- pattern_match(Text, loud_line), pattern_match(Text, inquiry_line), !.
hey(Text, "Sure.") :- pattern_match(Text, inquiry_line), !.
hey(Text, "Whoa, chill out!") :- pattern_match(Text, loud_line), !.
hey(_, "Whatever.") :- !.

File: solutions/javascript/bob/1/bob.js
//
// For my Shree DR.MDD
//

export const hey = (input) => {
  const cleanedInput = input.trim();
  if (!cleanedInput) {
    return responses['response4'];
  }

  const shouting = isAllCaps(cleanedInput);
  const questioning = cleanedInput.endsWith('?');

  if (shouting && questioning) {
    return responses['response3'];
  } else if (questioning) {
    return responses['response1'];
  } else if (shouting) {
    return responses['response2'];
  } else {
    return responses['response5'];
  }
};

const responses = {
  'response1': 'Sure.',
  'response2': 'Whoa, chill out!',
  'response3': "Calm down, I know what I'm doing!",
  'response4': 'Fine. Be that way!',
  'response5': 'Whatever.',
};

function isAllCaps(text) {
  const lettersOnly = text.replace(/[^a-zA-Z]/g, '');
  return lettersOnly.length > 0 && lettersOnly === lettersOnly.toUpperCase();
}
